### PR TITLE
Vehicle window

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Collaboration session. 

Ps. I hate macros.

### CODE REVIEW

File: src/openloco/windows/vehicle.cpp [hunk 1]
```diff
@@ -622,13 +624,122 @@ namespace openloco::ui::vehicle
             common::sub_4B5ECD(self);
         }
 
+        struct VehicleStatus
+        {
+            uint16_t bx;
+            uint32_t eax;
+            uint16_t cx;
```
Comment (duncanspumpkin) [inline src/openloco/windows/vehicle.cpp:631]: `string_id` for waiting at or velocity.

File: src/openloco/windows/vehicle.cpp [hunk 2]
```diff
@@ -622,13 +624,122 @@ namespace openloco::ui::vehicle
             common::sub_4B5ECD(self);
         }
 
+        struct VehicleStatus
+        {
+            uint16_t bx;
```
Comment (duncanspumpkin) [inline src/openloco/windows/vehicle.cpp:629]: `string_id` for the main status message

File: src/openloco/windows/vehicle.cpp [hunk 3]
```diff
@@ -622,13 +624,122 @@ namespace openloco::ui::vehicle
...
+            uint32_t eax;
```
Comment (duncanspumpkin) [inline src/openloco/windows/vehicle.cpp:630]: eax and edx are the arguments for the messages.

File: src/openloco/windows/vehicle.cpp [hunk 4]
```diff
@@ -622,13 +624,122 @@ namespace openloco::ui::vehicle
...
+
+            auto veh = common::getVehicle(self);
+            auto status = sub_4B671C(veh);
+            FormatArguments args = {};
+            args.push(status.bx);
+            args.push(status.eax);
+            args.push(status.cx);
+            args.push(status.edx);
+
+            string_id bx = 456;
```
Comment (duncanspumpkin) [inline src/openloco/windows/vehicle.cpp:677]: 2 string status

File: src/openloco/objects/interface_skin_object.h [hunk 5]
```diff
@@ -184,12 +185,20 @@ namespace openloco
         constexpr uint32_t toolbar_menu_towns = 189;
         constexpr uint32_t toolbar_menu_stations = 190;
         constexpr uint32_t toolbar_menu_industries = 191;
+        constexpr uint32_t spr_192 = 192;
```
Comment (AaronVanGeffen) [inline src/openloco/objects/interface_skin_object.h:188]: Route?

**Comment by duncanspumpkin** [comment]
I've rebased this but plan to rework history so that the vehicle changes are in the correct commits

**Comment by duncanspumpkin** [comment]
Rebased and now compiling. Vehicle refactor now incorporated in correct commits.

File: src/openloco/localisation/string_ids.h [hunk 6]
```diff
@@ -1316,4 +1352,24 @@ namespace openloco::string_ids
 
     constexpr string_id menu_quit_to_menu = 2139;
     constexpr string_id menu_exit_openloco = 2140;
+
+    constexpr string_id getVehicleType(VehicleType type)
```
Comment (AaronVanGeffen) [inline src/openloco/localisation/string_ids.h:1356]: I think this function should be moved elsewhere. Perhaps a vehicle-related file?

**Comment by duncanspumpkin** [comment]
Rename length_of_var4C to orderTableOffset. Rename var4C to sizeOfOrderTable

File: src/openloco/game_commands.h [hunk 7]
```diff
@@ -226,11 +226,12 @@ namespace openloco::game_commands
...
     }
 
-    inline bool do_35(thing_id_t head, uint8_t order, uint16_t orderArgument, uint32_t orderOffset)
+    inline bool do_35(thing_id_t head, uint8_t order, uint64_t orderArgument, uint32_t orderOffset)
     {
         registers regs;
         regs.bl = GameCommandFlag::apply;
-        regs.ax = order | (orderArgument << 3);
+        regs.eax = order | (orderArgument << 3);
+        regs.cx = orderArgument >> 32;
```
Comment (AaronVanGeffen) [inline src/openloco/game_commands.h:234]: `cx` is only 16 bits. Does this need to be masked, explicitly?

Comment (duncanspumpkin) [reply src/openloco/game_commands.h:234]: I don't think so. I'm planning on refactoring this when I've finished all of the route tab as its excessive as it is copying in 48bits when vast majority of the time it should be 8bits.

File: src/openloco/windows/vehicle.cpp [hunk 8]
```diff
@@ -1006,6 +1016,126 @@ namespace openloco::ui::vehicle
...
+
+            switch (head->mode)
+            {
+                case TransportMode::rail:
+                {
+                    auto trackObj = objectmgr::get<track_object>(head->track_type);
+                    pickupImage = trackObj->var_1E + (pickupDisabled ? 17 : 16);
+                    pickupTooltip = pickupDisabled ? string_ids::tooltip_place_on_track : string_ids::tooltip_remove_from_track;
+                }
+                break;
```
Comment (AaronVanGeffen) [inline src/openloco/windows/vehicle.cpp:1104]: Stylistically, I really prefer the `break`s to be _in_ the scope.

File: src/openloco/graphics/image_ids.h [hunk 9]
```diff
@@ -53,7 +53,8 @@ namespace openloco::image_ids
     constexpr uint32_t rotate_object = 2365;
 
     constexpr uint32_t stop_start = 2369;
-
```
Comment (AaronVanGeffen) [inline src/openloco/graphics/image_ids.h]: This whitespace probably shouldn't be removed, as the constants aren't consecutive (yet).

File: src/openloco/GameCommands.h [hunk 10]
```diff
@@ -35,6 +47,45 @@ namespace openloco::game_commands
     uint32_t doCommand(int esi, const registers& registers);
     bool sub_431E6A(const company_id_t company, map::tile_element* const tile = nullptr);
 
+    inline void do_0(thing_id_t source, thing_id_t dest)
+    {
+        registers regs;
+        regs.bl = GameCommandFlag::apply;
+        regs.dx = source;
+        regs.di = dest;
+        doCommand(static_cast<int32_t>(GameCommand::vehicle_rearange), regs);
```
Comment (AaronVanGeffen) [inline src/openloco/GameCommands.h:56]: There's a typo in the enum value used here:
```suggestion
        doCommand(static_cast<int32_t>(GameCommand::vehicle_rearrange), regs);
```

File: src/OpenLoco/Things/Vehicle.h [hunk 11]
```diff
@@ -15,13 +16,37 @@ namespace openloco
...
+
         namespace flags_38
         {
             constexpr uint8_t unk_0 = 1 << 0;
             constexpr uint8_t unk_1 = 1 << 1;
             constexpr uint8_t unk_3 = 1 << 3;
             constexpr uint8_t unk_4 = 1 << 4;
         }
+
+        enum class Status : uint8_t
```
Comment (duncanspumpkin) [inline src/OpenLoco/Things/Vehicle.h:33]: define a new name for this type due to Xcode leaking symbols.

File: src/OpenLoco/Ui.h [hunk 12]
```diff
@@ -41,6 +41,9 @@ namespace OpenLoco::Ui
         hand_pointer,
         busy,
         diagonal_arrows,
+        unk_25,
+        unk_26,
```
Comment (marijnvdwerf) [inline src/OpenLoco/Ui.h:45]:
```suggestion
        unk_25 = 25,
        unk_26 = 26,
```

File: src/OpenLoco/Windows/Vehicle.cpp [hunk 13]
```diff
@@ -1,177 +1,3547 @@
...
+                tooltipContent = car->front->id;
+            }
+            if (self->row_hover != tooltipContent)
+            {
+                self->row_hover = tooltipContent;
+                self->invalidate();
+            }
+
+            char* buffer = const_cast<char*>(StringManager::getString(StringIds::buffer_337));
+            if (*buffer != '\0')
```
Comment (marijnvdwerf) [inline src/OpenLoco/Windows/Vehicle.cpp:1115]:
```suggestion
            if (strlen(buffer) != 0)
```

File: src/OpenLoco/Windows/Vehicle.cpp [hunk 14]
```diff
@@ -1,177 +1,3547 @@
...
+        // 0x004B3880 TODO: common across 3 tabs
+        static void tooltip(FormatArguments& args, Ui::window* const self, const widget_index widgetIndex)
+        {
+            args.push(StringIds::tooltip_scroll_vehicle_list);
+
+            auto vehicle = Common::getVehicle(self);
+            args.push(StringIds::getVehicleType(vehicle->vehicleType));
+        }
+
+        // 0x004B3B3F
```
Comment (marijnvdwerf) [inline src/OpenLoco/Windows/Vehicle.cpp:1198]:
```suggestion
        // 0x004B3B18
```

**Review by marijnvdwerf — CHANGES_REQUESTED** [review]
Some stuff found so far

File: src/OpenLoco/Localisation/StringIds.h [hunk 15]
```diff
@@ -697,9 +748,16 @@ namespace OpenLoco::StringIds
     constexpr string_id num_trams_plural = 1273;
     constexpr string_id num_aircrafts_plural = 1274;
     constexpr string_id num_ships_plural = 1275;
+    constexpr string_id train = 1276;
+    constexpr string_id bus = 1277;
+    constexpr string_id truck = 1278;
+    constexpr string_id tram = 1279;
+    constexpr string_id aircraft = 1280;
+    constexpr string_id ship = 1281;
+    constexpr string_id str_1282 = 1282;
```
Comment (marijnvdwerf) [inline src/OpenLoco/Localisation/StringIds.h:757]: unnamed

File: src/OpenLoco/Windows/Vehicle.cpp [hunk 16]
```diff
@@ -1,177 +1,3547 @@
...
+            if (self->widgets != widgets)
+            {
+                self->widgets = widgets;
+                self->initScrollWidgets();
+            }
+
+            Common::setActiveTabs(self);
+
+            auto* headVehicle = Common::getVehicle(self);
+            auto ax = headVehicle->var_44;
```
Comment (marijnvdwerf) [inline src/OpenLoco/Windows/Vehicle.cpp:1566]: You could probably skip these intermediate variables

File: src/OpenLoco/Windows/Vehicle.cpp [hunk 17]
```diff
@@ -1,177 +1,3547 @@
...
+            auto primaryCargoId = Utility::bitScanForward(vehicleObject->primary_cargo_types);
+
+            int32_t index = 0;
+            for (uint16_t cargoId = 0; cargoId < ObjectManager::getMaxObjects(object_type::cargo); cargoId++)
+            {
+                auto cargoObject = ObjectManager::get<cargo_object>(cargoId);
+                if (cargoObject == nullptr)
+                    continue;
+
+                if ((cargoObject->var_12 & 2) == 0)
```
Comment (marijnvdwerf) [inline src/OpenLoco/Windows/Vehicle.cpp:1794]: Add constant for flag

File: src/OpenLoco/Objects/CargoObject.h [hunk 18]
```diff
@@ -7,12 +7,19 @@ namespace OpenLoco
...
-        uint32_t unit_inline_sprite;    // 0x0C
+        string_id name; // 0x0
+        uint16_t var_2;
+        uint16_t var_4;
+        string_id units_and_cargo_name; // 0x6
+        string_id unit_name_singular;   // 0x8
+        string_id unit_name_plural;     // 0xA
+        uint32_t unit_inline_sprite;    // 0xC
+        std::uint8_t pad_10[0x12 - 0x10];
+        uint8_t var_12;
```
Comment (marijnvdwerf) [inline src/OpenLoco/Objects/CargoObject.h:18]:
```suggestion
        uint8_t flags;
```

File: src/OpenLoco/Windows/Vehicle.cpp (hunk 16 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Windows/Vehicle.cpp:1566]: I'm going to go through and change all of these in a latter pr

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**